### PR TITLE
fix(persist): merge function not working with async storages

### PR DESCRIPTION
**Pull request by AnatoleLucet**
Context: https://github.com/pmndrs/zustand/pull/468#issuecomment-869501387

### CODE REVIEW

File: src/middleware.ts [hunk 1]
```diff
@@ -368,22 +368,23 @@ export const persist =
...
         }
-      })
-      .then(() => {
+
         stateFromStorageInSync = merge(
           stateFromStorageInSync as S,
           configResult
         )
+
+        set(stateFromStorageInSync as S)
```
Comment (dai-shi) [inline src/middleware.ts:384]:
```suggestion
        set(stateFromStorageInSync as S, true)
```
I feel like we should replace it, but i'm not so sure.
btw, can we call `set` in sync?

File: src/middleware.ts [hunk 2]
```diff
@@ -368,22 +368,23 @@ export const persist =
...
+        return setItem()
+      })
+      .then(() => {
         postRehydrationCallback?.(stateFromStorageInSync, undefined)
```
Comment (dai-shi) [inline src/middleware.ts:388]:
```suggestion
        postRehydrationCallback?.(stateFromStorage, undefined)
```
Now, as it' used for both sync/async, maybe rename it?

File: src/middleware.ts [hunk 3]
```diff
@@ -368,22 +368,23 @@ export const persist =
             )
           } else {
             stateFromStorageInSync = deserializedStorageValue.state
```
Comment (dai-shi) [inline src/middleware.ts:370]:
```suggestion
            return deserializedStorageValue.state
```
Now, it seems we can simplify this and remove L375 if-statement.

File: src/middleware.ts (hunk 1 above)
Comment (dai-shi) [reply src/middleware.ts:384]: hm, it was already called in the previous code too.

**Review by dai-shi — COMMENTED** [review]
overall looks good. seems there's some room for refactoring. what do you think?

File: src/middleware.ts (hunk 1 above)
Comment (AnatoleLucet) [reply src/middleware.ts:384]: the set in sync doesn't seem to trigger a render if that's what you're worried about

File: src/middleware.ts (hunk 2 above)
Comment (AnatoleLucet) [reply src/middleware.ts:388]: good idea :+1:

**Comment by AnatoleLucet** [comment]
Pushed some changes, let me know what you think

**Review by dai-shi — APPROVED** [review]
Looks good. Unless we have any failing tests, we can't notice possible bugs anyway. Let's merge this. Thanks for your work.

File: src/middleware.ts (hunk 1 above)
Comment (dai-shi) [reply src/middleware.ts]: if it doesn't error, it's fine. I mistook it was only for async.

**Pull request merged**